zubhub_backend/zubhub/projects/utils.py

```diff
@@ -310,6 +312,76 @@ def get_published_projects_for_user(user, all):
     return public.union(authenticated).union(visible_to).order_by("-created_on")
 
 
+def recommend_projects_by_likes(projects, project=None):
+    """
+    Appends most popular projects by likes to an input list, which may be empty
+
+    Params:
+    projects - may be empty or contain 2 or less projects
+    project - original project if it exists, None if original project does not exist
+
+    Returns list containing three projects
+    """
+
+    """ if three projects still not found, add from most popular by likes """
+    Project = apps.get_model('projects.Project')
+
+    for l in Project.objects.all().order_by('-likes_count'):
+        if len(projects) == 3:
+            return projects
+        if not l in projects and l != project:
+            projects.append(l)
+
+    return projects
+
+def recommend_projects(project):
+    """
+    Params:
+    project - Project object
+
+    Returns list of three projects to recommend to a user 
+    """
+    Project = apps.get_model('projects.Project')
+
+    # get project category
+    category = getattr(project, 'category')
+
+    # get project tags
+    tags = list(project.tags.all())
+
+    projects = list(Project.objects.none())
+
+    if len(tags) != 0 and category != None:
+        # loop through each tag of the original project
+        for i in tags:
+            # loop through each project in the same category in random order and check if it contains this tag
+            for j in Project.objects.filter(category__name=category).order_by('?'):
+                tags_j = list(j.tags.all())
+                # skip original project and any projects already added to projects
+                if j == project or j in projects:
+                    continue
+                # if the jth project's tags include the ith tag, add to projects
+                if i in tags_j:
+                    projects.append(j)
```

Comment:
In order to make this a little cleaner and prevent redundant checks, we can get all the projects that are `PUBLIC` and that aren't the project that was passed into the function with a filter query at the top. Then all subsequent filters can be done on that. This would remove the need for the 361 check and also ensure we aren't looking at projects that are drafts. Also we can add a filter to 358 that also ensures that the projects have tags that include the `ith` tag.  Check [this](https://docs.djangoproject.com/en/dev/ref/models/querysets/#in) for more info. This would also remove the need for the 364 line.